system/jlib/jtrace.cpp

```diff
@@ -69,3 +100,688 @@ const char* LogTrace::queryLocalId() const
+
+    //local nested span parent
+    CSpan * localParentSpan = nullptr;
+};
+
+class CServerSpan : public CSpan
+{
+private:
+    void setSpanContext(StringArray & httpHeaders, const char kvDelineator = ':')
+    {
+        Owned<IProperties> contextProps = createProperties();
+        ForEachItemIn(currentHeaderIndex, httpHeaders)
+        {
+            const char* httpHeader = httpHeaders.item(currentHeaderIndex);
+            if(!httpHeader)
+                continue;
+
+            const char* delineator = strchr(httpHeader, kvDelineator);
+            if(delineator == nullptr)
+                continue;
+
+            StringBuffer key, value;
+            key.append(delineator - httpHeader, httpHeader);
+            value.set(delineator + 1);
+
+            contextProps->setProp(key, value);
+        }
+
+        setSpanContext(contextProps);
+    }
+
+    void setSpanContext(const IProperties * httpHeaders)
+    {
+        if (httpHeaders)
+        {
+            // perform any key mapping needed...
+            //Instrumented http client/server Capitalizes the first letter of the header name
+            //if (key == opentel_trace::propagation::kTraceParent || key == opentel_trace::propagation::kTraceState )
+            //    theKey[0] = toupper(theKey[0]);
+
+            if (httpHeaders->hasProp(HPCCSemanticConventions::kGLOBALIDHTTPHeader))
+                hpccGlobalId.set(httpHeaders->queryProp(HPCCSemanticConventions::kGLOBALIDHTTPHeader));
+            else
+                DBGLOG("ServerSpan: HPCCGlobalID not found in http headers");
+
+            if (httpHeaders->hasProp(HPCCSemanticConventions::kCallerIdHTTPHeader))
+                hpccCallerId.set(httpHeaders->queryProp(HPCCSemanticConventions::kCallerIdHTTPHeader));
+            else
+            {
+                DBGLOG("ServerSpan: HPCCCallerID not provied");
+            }
+
+
+            const CHPCCHttpTextMapCarrier carrier(httpHeaders);
+            auto globalPropegator = context::propagation::GlobalTextMapPropagator::GetGlobalPropagator();
+            //avoiding getCurrent: https://github.com/open-telemetry/opentelemetry-cpp/issues/1467
+            opentelemetry::context::Context empty;
+            auto wrapped_span_context = SetSpan(empty, span);
+            opentelemetry::v1::context::Context runtimeCtx = globalPropegator->Extract(carrier, wrapped_span_context);
+            opentelemetry::v1::nostd::shared_ptr<opentelemetry::v1::trace::Span> remoteParentSpan = opentelemetry::trace::GetSpan(runtimeCtx);
+
+            if (remoteParentSpan != nullptr && remoteParentSpan->GetContext().IsValid())
+            {
+                remoteParentSpanCtx = remoteParentSpan->GetContext();
+                opts.parent = remoteParentSpanCtx;
+            }
+            else
+                WARNLOG("ServerSpan: Could not create remote parent span based on OTel parenttrace/tracestate headers");
+        }
+
+        //if (!httpHeaders || !parentContext.IsValid())
```

Comment:
If a server span does not receive otel or hpcc context, do we create new ones?

Reply:
In general, yes, although we may want to it to be configured by an option.  I suspect we should *always* preserve ids and pass them through, but it may be an option whether we create one.  I would be tempted to default to always creating them, and add an option if there is a reason to turn it off.

Reply:
> In general, yes, although we may want to it to be configured by an option. I suspect we should _always_ preserve ids and pass them through, but it may be an option whether we create one. I would be tempted to default to always creating them, and add an option if there is a reason to turn it off.

Definitely on the same page regarding the preservation and propagation of IDs (Otel and HPCC).

Agreed on defaulting to generate IDs as well, and exposing option to suppress new span generation (this might have to wait until later phases).

What I'm not totally clear on is...
Is there any benefit in generating IDs for the remote parent/caller if we don't receive or can not process incoming tracing context.
We should discuss.